gen/top_gen.ml

```diff
@@ -432,69 +443,88 @@ let max_set = IntSet.max_elt
       if (not (StringSet.mem loc atoms) && O.optcond) then k
       else cons_one loc v k in
 
-    let rec check_rec p i =
-
+    (* - `p`, process number
+       - `i`, initial value accumulator
+       - input `xvs`, type `U.cos` (defined in topUtil.ml)
+         the final values of write events for all locations *)
+    let check p i xvs =
       let open Config in
-      function
-      | [] -> i,[],[]
-      (* `vs` : (IntSet.elt array * IntSet.t) list list *)
-      | (x,vs)::xvs ->
-         let vs = List.map ( List.map ( fun (v, vset) -> (Array.map Code.value_to_int v, vset) ) ) vs in
-         let i,c,f = match O.cond with
-           | Observe ->
-              let vs = List.flatten vs in
-              begin match vs with
-              | [] -> i,[],[]
-              | _::_ ->
-                  let v,_ = Misc.last vs in
-                  i,[],cons_one x v []
-              end
+      (* The accumulator:
+        - `p` procedure number.
+        - `i` initial value, of type `init` (defined in archExtra_gen.ml).
+          It remains unchanged in the default configuration.
+          It is only updated via `call_build_observers`.
+        - `cs` pseudo code. It is empty in the default configuration.
+          It is only updated via `call_build_observers`.
+        Element of `vxs` is `(x, vs)`.
+        - `x` is the location represented by a string
+        - `vs` the final value of the location `x` *)
+      let _p,i,cs,fs = List.fold_left (fun (p, i, cs, fs) (x, vs) ->
+        (* `vs` : (IntSet.elt array * IntSet.t) list list *)
```

Comment:
Any reason why this type comment can't be an actual type annotation? That would give the extra advantage that the compile can check that the type annotation is actually correct and doesn't go out of sync with the code. In this case, for example, it seems like the type of `vs` is actually `(Code.v array * IntSet.t) list list` rather than the type in the comment.